File: Python/PE081.py
##Problem 81
##
##In the 5 by 5 matrix below, the minimal path sum from the top left to the bottom right, by only moving to the right and down, is indicated in bold red and is equal to 2427.
##
##Find the minimal path sum, in matrix.txt (right click and "Save Link/Target As..."), a 31K text file containing a 80 by 80 matrix, from the top left to the bottom right by only moving right and down.


fichier=open('PE081_matrix.txt','r')
L=fichier.readlines()
fichier.close()

for i, l in enumerate(L):
    L[i] = [int(m) for m in l.split(',')]

b = len(L[0]) #base de numérotation et largeur de la matrice
h = len(L) # hauteur de la matrice

L_u = {} #liste des noeuds non visités
en_cours = {}
traites = {}

en_cours['D'] = [0, [[0,L[0][0]]]]

infini = 80*80*9999

for i, l in enumerate(L):
    for j, c in enumerate(l):
        voisins = []
        if i<h-1:
            voisins.append([(i+1)*b+j, L[i+1][j]])
        if j<b-1:
            voisins.append([i*b+j+1, L[i][j+1]])
        L_u[i*b+j] = [infini, voisins]



while len(en_cours)>0:
    distMini = infini
    clefMini = -1
    #1 recherche du plus proche noeud
    for n in en_cours:
        if en_cours[n][0] < distMini:
            distMini = en_cours[n][0]
            clefMini = n

    current = en_cours.pop(clefMini)

    #2 balayge de ses voisins
    for (num, dist) in current[1]:
        if num in L_u:
            en_cours[num] = L_u.pop(num)

        if num in en_cours:
            en_cours[num][0] = min(en_cours[num][0], current[0] + dist)

    #3 stockage du courant :
    traites[clefMini] = current

print(traites[80*79+79])
            

File: Lua/pe081.lua
--[[
PE081.lua

In the 5 by 5 matrix below, the minimal path sum from the top left to the bottom
right, by only moving to the right and down, is indicated in bold red and is
equal to 2427.


131	673	234	103	18
201	96	342	965	150
630	803	746	422	111
537	699	497	121	956
805	732	524	37	331


Find the minimal path sum, in matrix.txt (right click and
'Save Link/Target As...'), a 31K text file containing a 80 by 80 matrix, from
the top left to the bottom right by only moving right and down.

--]]

DIM=80


-- lecture du fichier et stockage dans fichier{}
io.input("pe081_matrix.txt",r)
line=nil
fich={}
i=1
for line in io.lines() do
  fich[i]=line
  i=i+1
end
io.close()


for i=1,table.maxn(fich) do
  local lgLine=string.len(fich[i])
  local t={}
  local c_s=1
  local c_e=string.find(fich[i],',') --,c_s)
  local j=1
  while c_e>c_s do
    t[j]=tonumber(string.sub(fich[i],c_s,c_e-1))
    c_s=c_e+1
    c_e=string.find(fich[i],',',c_s) or lgLine+1
    j=j+1
  end
  fich[i]=t
end

--fich=nil
--fich={{131,673,234,103,18},{201,96,342,965,150},{630,803,746,422,111},{537,699,497,121,956},{805,732,524,37,331}}

res={}
for i=1,DIM do
  local t={}
  res[i]=t
end

for i=1,DIM do -- ligne
  for j=1,DIM do -- colonne
    --print(i,j,fich[i][j])
	if i==1 then
	  res[i][j]=(res[i][j-1] or 0)+fich[i][j]
	elseif j==1 then
	  res[i][j]=(res[i-1][j] or 0)+fich[i][j]
	else
      res[i][j]=math.min(res[i][j-1],res[i-1][j])+fich[i][j]
	end
  end
end
print("-->",res[DIM][DIM])

